### PR TITLE
Examples 

### PR DESCRIPTION
**Pull request by Vincent26-Chen**
@vpaprots

### CODE REVIEW

File: golang/examples/server_test.go [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+
+	"github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/ep11"
+	pb "github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/grpc"
+	"github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/util"
+	uuid "github.com/satori/go.uuid"
+	grpc "google.golang.org/grpc"
+)
+
+const (
+	address = "zlxcn002.torolab.ibm.com:29876"
```
Comment (vpaprots) [inline golang/examples/server_test.go]: lets not leak this hostname externally

File: .gitignore [hunk 2]
```diff
@@ -1 +1,6 @@
-golang/examples/examples
+.gitignore
```
Comment (vpaprots) [inline .gitignore]:
```
# IDEA IDE
.idea
.DS_Store

# Visual Studio IDE
.vscode
```

File: README.md [hunk 3]
```diff
@@ -4,4 +4,15 @@ This repo contains software to be used to connect to the cloud service **IBM Clo
 
 # Contents
 
-The contents of this repo are offered *as-is* and is subject to be changed anytime. This is an *Experimental Service* from IBM Cloud at this point in time. More info about Terms and Conditions: http://www-03.ibm.com/software/sla/sladb.nsf/sla/bm-6605-12
+The contents of this repo are offered *as-is* and is subject to change at anytime. This is an *Experimental Service* from IBM Cloud at this point in time. To see more information about the Terms and Conditions, review http://www-03.ibm.com/software/sla/sladb.nsf/sla/bm-6605-12.
+
+# Golang Example
```
Comment (vpaprots) [inline README.md:9]: Golang Examples

File: README.md [hunk 4]
```diff
@@ -4,4 +4,15 @@ This repo contains software to be used to connect to the cloud service **IBM Clo
...
+
+The Golang examples can be found [here](golang/README.md). The examples show how to manipulate with **IBM Cloud Hyper Protect Crypto Services** to accomplish following functions:
+
+* Get mechenism information
+* Encrypt and decrypt
+* Sign and verify
+* Wrap and unwrap key
+* Derive key
+* Build message digest
```
Comment (vpaprots) [inline README.md:18]: Message digest

File: README.md [hunk 5]
```diff
@@ -4,4 +4,15 @@ This repo contains software to be used to connect to the cloud service **IBM Clo
...
 # Contents
 
-The contents of this repo are offered *as-is* and is subject to be changed anytime. This is an *Experimental Service* from IBM Cloud at this point in time. More info about Terms and Conditions: http://www-03.ibm.com/software/sla/sladb.nsf/sla/bm-6605-12
+The contents of this repo are offered *as-is* and is subject to change at anytime. This is an *Experimental Service* from IBM Cloud at this point in time. To see more information about the Terms and Conditions, review http://www-03.ibm.com/software/sla/sladb.nsf/sla/bm-6605-12.
+
+# Golang Example
+
+The Golang examples can be found [here](golang/README.md). The examples show how to manipulate with **IBM Cloud Hyper Protect Crypto Services** to accomplish following functions:
+
+* Get mechenism information
```
Comment (vpaprots) [inline README.md:13]: EP11 Mechanism Info

File: README.md (hunk 4 above)
Comment (vpaprots) [reply README.md:18]: * TLS Signer example

File: golang/README.md [hunk 6]
```diff
@@ -1,23 +1,33 @@
 # Overview
 
-- [Install golang](https://golang.org/doc/install)
-- [Change GREP11 server address](https://github.com/vpaprots/ibm-cloud-hyperprotectcrypto/blob/master/golang/examples/main.go#L146)
+1. [Install Golang](https://golang.org/doc/install)
+2. [Change GREP11 server address](https://github.com/Vincent26-Chen/ibm-cloud-hyperprotectcrypto/blob/master/golang/examples/server_test.go#L22) in `examples/server_test.go` file
```
Comment (vpaprots) [inline golang/README.md]: Lets use local links if possible `[Change GREP11 server address](golang/examples/main.go)`

File: golang/README.md [hunk 7]
```diff
@@ -1,23 +1,33 @@
...
+    === RUN   Example_encryptAndecrypt
+    --- PASS: Example_encryptAndecrypt (0.05s)
+    === RUN   Example_digest
+    --- PASS: Example_digest (0.02s)
+    === RUN   Example_signAndVerifyUsingRSAKeyPair
+    --- PASS: Example_signAndVerifyUsingRSAKeyPair (0.65s)
+    === RUN   Example_wrapAndUnwrapKey
+    --- PASS: Example_wrapAndUnwrapKey (0.64s)
+    === RUN   Example_deriveKey
+    --- PASS: Example_deriveKey (0.11s)
```
Comment (vpaprots) [inline golang/README.md]: TLS?

File: golang/examples/server_test.go [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+	fmt.Printf("Got mechanism list:\n%v ...\n", mechanismListResponse.Mechs[:5])
+
+	mechanismInfoRequest := &pb.GetMechanismInfoRequest{
+		Mech: ep11.CKM_RSA_PKCS,
+	}
+	_, err = cryptoClient.GetMechanismInfo(context.Background(), mechanismInfoRequest)
+	if err != nil {
+		panic(fmt.Errorf("Get mechanism info error: %s", err))
+	}
+	fmt.Printf("Got CKM_RSA_PKCS mechanism info\n")
```
Comment (vpaprots) [inline golang/examples/server_test.go]: This looks extra

File: golang/examples/server_test.go [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+
+	// Output:
+	// Got mechanism list:
+	// [CKM_RSA_PKCS CKM_RSA_PKCS_KEY_PAIR_GEN CKM_RSA_X9_31_KEY_PAIR_GEN CKM_RSA_PKCS_PSS CKM_SHA1_RSA_X9_31] ...
+	// Got CKM_RSA_PKCS mechanism info
+}
+
+//Example_encryptAndecrypt encrypt and decrypt a piece of text.
```
Comment (vpaprots) [inline golang/examples/server_test.go:51]: Needs ascii art

File: golang/examples/server_test.go [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+	if err != nil {
+		panic(fmt.Errorf("did not connect: %v", err))
+	}
+	defer conn.Close()
+
+	cryptoClient := pb.NewCryptoClient(conn)
+
+	digestData := []byte("This is the data longer than 64 bytes so that multiple digest operations are needed")
+	digestInitRequest := &pb.DigestInitRequest{
+		Mech: &pb.Mechanism{Mechanism: ep11.CKM_SHA_1},
```
Comment (vpaprots) [inline golang/examples/server_test.go]: Lets use SHA256

File: golang/examples/server_test.go [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+		fmt.Printf("Digest data using multiple operations: %x\n", digestFinalResponse.Digest)
+	}
+
+	// Output:
+	// Digest data using single digest operation: 56aee95b369b72e16969b8d5f8fe940019afac98
+	// Digest data using multiple operations: 56aee95b369b72e16969b8d5f8fe940019afac98
+}
+
+// Example_signAndVerifyUsingRSAKeyPair sign on a piece of data and verify it
+func Example_signAndVerifyUsingRSAKeyPair() {
```
Comment (vpaprots) [inline golang/examples/server_test.go:252]: Lets add ECDSA

File: golang/examples/server_test.go [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+	conn, err := grpc.Dial(address, grpc.WithInsecure())
+	if err != nil {
+		panic(fmt.Errorf("did not connect: %v", err))
+	}
+	defer conn.Close()
+
+	cryptoClient := pb.NewCryptoClient(conn)
+
+	//Generate a DES3 key
+	desKeyTemplate := util.NewAttributeMap(
```
Comment (vpaprots) [inline golang/examples/server_test.go:524]: Lets use AES key

File: golang/examples/server_test.go [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+}
+
+var (
+	oidNamedCurveP224 = asn1.ObjectIdentifier{1, 3, 132, 0, 33}
+	oidNamedCurveP256 = asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7}
+	oidNamedCurveP384 = asn1.ObjectIdentifier{1, 3, 132, 0, 34}
+	oidNamedCurveP521 = asn1.ObjectIdentifier{1, 3, 132, 0, 35}
+)
+
+func namedCurveFromOID(oid asn1.ObjectIdentifier) elliptic.Curve {
```
Comment (vpaprots) [inline golang/examples/server_test.go]: There is a copy of this in utilities, lets merge

File: golang/examples/server_test.go [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+	KeyType asn1.ObjectIdentifier
+	Curve   asn1.ObjectIdentifier
+}
+
+type PubKeyASN struct {
+	Ident EckeyIdentASN
+	Point asn1.BitString
+}
+
+func blobToPubKey(pubKey []byte, curve asn1.ObjectIdentifier) ([]byte, *ecdsa.PublicKey, error) {
```
Comment (vpaprots) [inline golang/examples/server_test.go]: Ditto, exists in utilities

File: golang/examples/server_test.go [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+	_, pubKey, err := blobToPubKey(spki, oidNamedCurveP256)
+	if err != nil {
+		return nil, err
+	}
+	combinedCoordinates := elliptic.Marshal(namedCurveFromOID(oidNamedCurveP256), pubKey.X, pubKey.Y)
+	return combinedCoordinates, nil
+}
+
+//Example_deriveKey generates ECDH key pairs for Bob and Alice, then generates AES keys for both of them
+func Example_deriveKey() {
```
Comment (vpaprots) [inline golang/examples/server_test.go:615]: If you have a reference for this.. include a link

File: golang/examples/server_test.go [hunk 16]
```diff
@@ -0,0 +1,594 @@
...
+	x, y := elliptic.Unmarshal(nistCurve, decode.Point.Bytes)
+	if x == nil {
+		return nil, nil, fmt.Errorf("failed unmarshalling public key.\n%s", hex.Dump(decode.Point.Bytes))
+	}
+
+	return ski, &ecdsa.PublicKey{Curve: nistCurve, X: x, Y: y}, nil
+}
+
+//getCoordinatesByECSPKI extracts coordinates bit array from public key in SPKI format
+func getCoordinatesByECSPKI(spki []byte) ([]byte, error) {
```
Comment (vpaprots) [inline golang/examples/server_test.go]: part of utilities?

File: golang/examples/server_test.go [hunk 17]
```diff
@@ -0,0 +1,594 @@
...
+	if err != nil {
+		panic(fmt.Errorf("Bob EC Key Derive Error: %s", err))
+	}
+	if !bytes.Equal(aliceDerivekeyResponse.CheckSum, bobDerivekeyResponse.CheckSum) {
+		panic(fmt.Errorf("Alice and Bob Get Different Derive Key Checksum: %v -- %v",
+			aliceDerivekeyResponse.CheckSum, bobDerivekeyResponse.CheckSum))
+	} else {
+		fmt.Println("Alice and Bob get same derived key")
+	}
+
```
Comment (vpaprots) [inline golang/examples/server_test.go:726]: EncryptSingle with aliceDerivekeyResponse.Key and decryptsingle with bobDerivekeyResponse.Key

File: golang/examples/tls_test.go [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+	resp.Body.Close()
+	if err != nil {
+		return "", fmt.Errorf("ioutil.ReadAll failed: %s", err)
+	}
+	return string(data), nil
+}
+
+//Example_tls tests TLS communication between client and server, where certificate and private key are generated on the fly
+func Example_tls() {
+	conn, err := openGrep11Connection(address)
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: remove openGrep11Connection functions

File: golang/util/signer.go [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+
+	"github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/ep11"
+	pb "github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/grpc"
+)
+
+//EP11PrivateKey MUST implement crypto.Signer interface so that tls crypt/tls package can take as EP11PrivateKey in
+//tls.Certificate: https://golang.org/pkg/crypto/tls/#Certificate
+type EP11PrivateKey struct {
+	algorithmOID asn1.ObjectIdentifier
+	keyBlob      *[]byte
```
Comment (vpaprots) [inline golang/util/signer.go]: not a pointer type

File: golang/util/signer.go [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+		return nil, fmt.Errorf("Unsupported Public key type: %v", priv.algorithmOID)
+	}
+}
+
+//Public implement crypto.Signer interface
+func (priv *EP11PrivateKey) Public() crypto.PublicKey {
+	return priv.pubKey
+}
+
+func NewEP11Signer(cryptoClient pb.CryptoClient, privKeyBlob *[]byte, spki *[]byte) (*EP11PrivateKey, error) {
```
Comment (vpaprots) [inline golang/util/signer.go]: no pointers for slices

File: golang/util/signer.go [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+	priv := &EP11PrivateKey{}
```
Comment (vpaprots) [inline golang/util/signer.go]: priv := &EP11PrivateKey{
   cryptoClient: cryptoClient
....
}

File: golang/util/util.go [hunk 22]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
 
 	return false, err2
 }
+
+var (
+	oidNamedCurveP224 = asn1.ObjectIdentifier{1, 3, 132, 0, 33}
```
Comment (vpaprots) [inline golang/util/util.go]: Lets capitalize

File: golang/util/util.go [hunk 23]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
...
+	oidNamedCurveP256 = asn1.ObjectIdentifier{1, 2, 840, 10045, 3, 1, 7} //prime256v1
+	oidNamedCurveP384 = asn1.ObjectIdentifier{1, 3, 132, 0, 34}
+	oidNamedCurveP521 = asn1.ObjectIdentifier{1, 3, 132, 0, 35}
+	oidECPublicKey    = asn1.ObjectIdentifier{1, 2, 840, 10045, 2, 1}
+	oidRSAPublicKey   = asn1.ObjectIdentifier{1, 2, 840, 113549, 1, 1, 1}
+)
+
+//GetNamedCurveFromOID returns Curve from specified curve OID
+func getNamedCurveFromOID(oid asn1.ObjectIdentifier) elliptic.Curve {
```
Comment (vpaprots) [inline golang/util/util.go]: Can make this one public

File: golang/util/util.go [hunk 24]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
...
+	case oid.Equal(oidNamedCurveP384):
+		return elliptic.P384()
+	case oid.Equal(oidNamedCurveP521):
+		return elliptic.P521()
+	}
+	return nil
+}
+
+//GetNamedCurveAndOIDFromNumBits returns EC curve && curve OID supported by elliptic package
+func getNamedCurveAndOIDFromNumBits(nbits int32) (elliptic.Curve, asn1.ObjectIdentifier) {
```
Comment (vpaprots) [inline golang/util/util.go]: Remove

File: golang/util/util.go [hunk 25]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
...
+	case 384:
+		return elliptic.P384(), oidNamedCurveP384
+	case 521:
+		return elliptic.P521(), oidNamedCurveP521
+	}
+	return nil, nil
+}
+
+//GetCurveOIDFromNumBits returns curve OID from number of bits.
+func GetCurveOIDFromNumBits(nbits int32) asn1.ObjectIdentifier {
```
Comment (vpaprots) [inline golang/util/util.go]: remove

File: golang/examples/tls_test.go [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+func openGrep11Connection(address string) (*grpc.ClientConn, error) {
+	conn, err := grpc.Dial(address, grpc.WithInsecure())
+	if err != nil {
+		return nil, fmt.Errorf("Cannot connect: %s", err)
+	}
+	return conn, nil
+}
+
+//generateECDSAKeyPair generate ECDSA key pair
+func generateECDSAKeyPair(cryptoClient pb.CryptoClient, nBits int32) ([]byte, []byte, error) {
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: no bits

File: golang/util/util.go [hunk 27]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
...
+	KeyType asn1.ObjectIdentifier
+}
+
+//generalPubKeyASN used to identify the public key type
+type generalPubKeyASN struct {
+	OIDAlgorithm pubKeyTypeASN
+}
+
+//GetPubKey convert ep11 SPKI structure to golang ecdsa.PublicKey
+func GetPubKey(spki *[]byte) (crypto.PublicKey, asn1.ObjectIdentifier, error) {
```
Comment (vpaprots) [inline golang/util/util.go]: no pointer

File: golang/util/util.go [hunk 28]
```diff
@@ -135,3 +139,120 @@ func Convert(err error) (bool, *pb.Grep11Error) {
...
+
+	} else if firstDecode.OIDAlgorithm.KeyType.Equal(oidRSAPublicKey) {
+		return nil, nil, fmt.Errorf("RSA public key not supported yet")
+	} else {
+		return nil, nil, fmt.Errorf("Unrecognized public key type %v", firstDecode.OIDAlgorithm)
+	}
+}
+
+//GetPubkeyBytesFromSPKIBlob extracts coordinates bit array from public key in SPKI format
+func GetPubkeyBytesFromSPKIBlob(spki []byte) ([]byte, error) {
```
Comment (vpaprots) [inline golang/util/util.go]: Remove?

File: golang/util/signer.go [hunk 29]
```diff
@@ -0,0 +1,81 @@
...
+//EP11PrivateKey MUST implement crypto.Signer interface so that tls crypt/tls package can take as EP11PrivateKey in
+//tls.Certificate: https://golang.org/pkg/crypto/tls/#Certificate
+type EP11PrivateKey struct {
+	algorithmOID asn1.ObjectIdentifier
+	keyBlob      *[]byte
+	pubKey       crypto.PublicKey //&ecdsa.PublicKey{} (rsa PublicKey not support yet)
+	cryptoClient pb.CryptoClient
+}
+
+type esDSASignature struct {
```
Comment (vpaprots) [inline golang/util/signer.go]: ecdsaSignature

Comment (vpaprots) [reply golang/util/signer.go]: move into Sign()

File: golang/examples/tls_test.go [hunk 30]
```diff
@@ -0,0 +1,211 @@
...
+	}
+	var ecKeypairResponse *pb.GenerateKeyPairResponse
+	ecKeypairResponse, err = cryptoClient.GenerateKeyPair(context.Background(), generateECKeypairRequest)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Generate Alice EC Key Pair Error: %s", err)
+	}
+	return ecKeypairResponse.PrivKey, ecKeypairResponse.PubKey, nil
+}
+
+func createECDSASelfSignedCert(privKey *util.EP11PrivateKey, commonName string,
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: fits into 180 chars on single lines?

File: golang/examples/tls_test.go [hunk 31]
```diff
@@ -0,0 +1,211 @@
...
+
+	certDERBytes, err := x509.CreateCertificate(rand.Reader, &template, &template, privKey.Public(), privKey)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create certificate: %s", err)
+	}
+	return certDERBytes, nil
+}
+
+//StartServer starts https server
+func CreateServer(tlsCert tls.Certificate) *http.Server {
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: lets use *tls.Config

File: golang/examples/tls_test.go [hunk 32]
```diff
@@ -0,0 +1,211 @@
...
+	tlsCert := tls.Certificate{
+		Certificate: [][]byte{certDER},
+		PrivateKey:  priv,
+	}
+
+	//start server thread
+	httpServer := CreateServer(tlsCert)
+	defer httpServer.Close()
+	go func() {
+		httpServer.ListenAndServeTLS("", "")
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: lis, err = tls.Listen("tcp", address, tlsConfig)
httpServer.Serve(lis)

File: golang/examples/tls_test.go [hunk 33]
```diff
@@ -0,0 +1,211 @@
...
+	}()
+
+	time.Sleep(1 * time.Second)
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: dont need this if have tls.Listen open

File: golang/examples/tls_test.go [hunk 34]
```diff
@@ -0,0 +1,211 @@
...
+	}
+	return httpServer
+}
+
+type httpTestClient struct {
+	httpClient *http.Client
+	serverAddr string
+}
+
+func newHTTPTestClient(caCertDER []byte, serverAddr string) *httpTestClient {
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: return *http.Client

File: golang/examples/tls_test.go [hunk 35]
```diff
@@ -0,0 +1,211 @@
...
+	defer httpServer.Close()
+	go func() {
+		httpServer.ListenAndServeTLS("", "")
+	}()
+
+	time.Sleep(1 * time.Second)
+
+	//create TLS client
+	client := newHTTPTestClient(certDER, httpServer.Addr)
+	strResp, err := client.ping()
```
Comment (vpaprots) [inline golang/examples/tls_test.go]: ping(Addr)

File: golang/examples/tls_test.go (hunk 31 above)
Comment (Vincent26-Chen) [reply golang/examples/tls_test.go]: If using tls.Listen(), tls.Config is not needed by http server anymore.

**Comment by Vincent26-Chen** [comment]
all changes from time of fork are squashed into one. @vpaprots

File: README.md (hunk 5 above)
Comment (vpaprots) [reply README.md:13]: John>

File: golang/README.md [hunk 36]
```diff
@@ -1,23 +1,45 @@
...
-GOPATH/src/github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/examples$ go run main.go
-Generated AES Key
-Generated IV
-Encrypted message
-Decrypted message
-Hello, this is a very long and creative message without any imagination
-GOPATH/src/github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/examples$ 
-```
+1. [Install Golang](https://golang.org/doc/install)
+2. [Change GREP11 server address](examples/server_test.go#L18) in `examples/server_test.go` file
```
Comment (vpaprots) [inline golang/README.md:4]: John> add comments about IAM auth

File: golang/README.md [hunk 37]
```diff
@@ -1,23 +1,45 @@
...
+	--- PASS: Example_deriveKey (0.11s)
+	=== RUN   Example_tls
+	--- PASS: Example_tls (0.05s)
+	PASS
+	ok  	github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/examples	1.667s
+    ```
+
+## General function call workflow
+
+Grep11 can do encrypt, decrypt, digest, sign and verify operations. For each operation, there are a series of functions. For example, It has `EncryptInit()`, `Encrypt()`, `EncryptUpdate()`, `EncryptFinal()` and `EncryptSingle()`. In general `XXXInit()` is used to initialize a operation; `XXXUpdate()` and `XXXFinal()` are used to perform mutiple function calls against long pieace of message; `XXXSingle()` is IBM ep11 extention to the standard PKCS#11 specification and used to perform a single call without `XXXInit()` call. The following graph script shows the three ways to calling sequence. The `XXX` could be `Encrypt`, `Decrypt`, `Digest`, `Sign` and `Verify`.
```
Comment (vpaprots) [inline golang/README.md:34]: John>

File: golang/README.md [hunk 38]
```diff
@@ -1,23 +1,45 @@
...
+![function work flow](func_workflow.png)
+
 
 ## Things to consider
 
-This example doesn't use TLS. However, this is just standard GRPC client library, TLS documented online elsewheres.
+This example does not use TLS. If you would like to experiment with TLS, review https://grpc.io/docs/guides/auth.html on how to implement TLS with GRPC.
```
Comment (vpaprots) [inline golang/README.md:40]: John> IAM & Auth does use TLS..

File: golang/examples/tls_test.go [hunk 39]
```diff
@@ -0,0 +1,187 @@
...
+	resp.Body.Close()
+	if err != nil {
+		return "", fmt.Errorf("ioutil.ReadAll failed: %s", err)
+	}
+	return string(data), nil
+}
+
+//Example_tls tests TLS communication between client and server, where certificate and private key are generated on the fly
+func Example_tls() {
+	conn, err := grpc.Dial(address, grpc.WithInsecure())
```
Comment (vpaprots) [inline golang/examples/tls_test.go:128]: John> IAM

File: golang/README.md [hunk 40]
```diff
@@ -1,23 +1,45 @@
...
-GOPATH/src/github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/examples$ 
-```
+1. [Install Golang](https://golang.org/doc/install)
+2. [Change GREP11 server address](examples/server_test.go#L18) in `examples/server_test.go` file
+3. cd `$GOPATH/src/github.com/ibm-developer/ibm-cloud-hyperprotectcrypto/golang/examples`
+4. Execute the example by issuing the command `go test -v`
+5. You will see similiar output as the following from the sample program:
+
+    ```Bash
+	=== RUN   Example_getMechnismInfo
```
Comment (vpaprots) [inline golang/README.md:10]: IAM&TLS setup example

File: golang/util/util.go [hunk 41]
```diff
@@ -135,3 +139,89 @@ func Convert(err error) (bool, *pb.Grep11Error) {
 
 	return false, err2
 }
+
```
Comment (vpaprots) [inline golang/util/util.go:142]: >>> `panic("PKCS11: Unhandled attribute type " + err.Error())` need to discuss

Comment (vpaprots) [reply golang/util/util.go:142]: Any examples of `Convert`? We need an example that does error handling

**Review by vpaprots — APPROVED** [review]
Some more comments for next PR

**Pull request merged**